code_generation/configuration.py

```diff
@@ -0,0 +1,526 @@
+        """
+        Function used to add a rule to the configuration.
+
+        Args:
+            scopes: The scopes to which the rule should be added. This can be a list of scopes or a single scope.
+            rule: The rule to be added. This must be a ProducerRule object.
+
+        Returns:
+            None
+
+        """
+        if not isinstance(rule, ProducerRule):
+            raise TypeError("Rule must be of type ProducerRule")
+        if not isinstance(scopes, list):
+            scopes = [scopes]
+        rule.set_scopes(scopes)
+        rule.set_global_scope(self.global_scope)
+        self.rules.add(rule)
+
+    def resolve_modifiers(self, configuration_dict: Dict[Any, Any]) -> TConfiguration:
+        """
+        Function used to resolve mofifiers used in the configuration. This function is called by the add_config_parameters function.
+
+        Args:
+            configuration_dict: The configuration dictionary to be resolved.
+
+        Returns:
+            dict: The resolved configuration dictionary.
+        """
+        resolved_dict: TConfiguration = {}
+        for key in configuration_dict:
+            resolved_value = None
+            log.debug("Resolved: {}".format(resolved_dict))
+            log.debug("Parameter: {}".format(key))
+            log.debug("Paramters: {}".format(configuration_dict))
+            if isinstance(configuration_dict[key], dict):
+                resolved_dict.update(self.resolve_modifiers(configuration_dict[key]))
+            elif isinstance(configuration_dict[key], SampleModifier):
+                log.debug("Applying Samplemodifier")
+                resolved_value = configuration_dict[key].apply(self.sample)
+            elif isinstance(configuration_dict[key], EraModifier):
+                log.debug("Applying Eramodifier")
+                resolved_value = configuration_dict[key].apply(self.era)
+            else:
+                resolved_value = configuration_dict[key]
+            resolved_dict.update({key: resolved_value})
+        return resolved_dict
+
+    def remove_empty_scopes(self) -> None:
+        """
+        Function used to remove empty scopes from the configuration. This function is called by the optimize function,
+        which should be called after all configuration changes have been made.
+
+        Args:
+            None
+
+        Returns:
+            None
+        """
+        scopes_to_test = [scope for scope in self.scopes]
+        for scope in scopes_to_test:
+            if (len(self.producers[scope]) == 0) or (
+                scope not in self.channels and scope is not self.global_scope
+            ):
+                log.warning("Removing unrequested / empty scope {}".format(scope))
+                self.scopes.remove(scope)
+                del self.producers[scope]
+                del self.outputs[scope]
+                del self.shifts[scope]
+                del self.config_parameters[scope]
```

Comment:
This might not be really necessary but I would also add 
`del self.available_outputs[scope]`
since it's in the loop
https://github.com/KIT-CMS/CROWN/blob/eff8756bcf070769078e2c7ed6d840eaac37f1e1/code_generation/configuration.py#L177-L182

Reply:
:+1: